storage/entity.go

```diff
@@ -0,0 +1,353 @@
+// Read more: https://docs.microsoft.com/rest/api/storageservices/fileservices/insert-or-replace-entity
+func (e *Entity) InsertOrReplace() error {
+	return e.insertOr(http.MethodPut)
+}
+
+// InsertOrMerge inserts an entity or merges the existing one.
+// Read more: https://docs.microsoft.com/en-us/rest/api/storageservices/fileservices/insert-or-merge-entity
+func (e *Entity) InsertOrMerge() error {
+	return e.insertOr("MERGE")
+}
+
+func (e *Entity) buildPath() string {
+	return fmt.Sprintf("%s(PartitionKey='%s', RowKey='%s')", e.Table.buildPath(), e.PartitionKey, e.RowKey)
+}
+
+// MarshalJSON is a custom marshaller for entity
+func (e *Entity) MarshalJSON() ([]byte, error) {
+	completeMap := map[string]interface{}{}
+	completeMap[partitionKeyNode] = e.PartitionKey
+	completeMap[rowKeyNode] = e.RowKey
+	for k, v := range e.Properties {
+		typeKey := strings.Join([]string{k, OdataTypeSuffix}, "")
+		switch t := v.(type) {
+		case []byte:
+			completeMap[typeKey] = OdataBinary
+			completeMap[k] = string(t)
+		case time.Time:
+			completeMap[typeKey] = OdataDateTime
+			completeMap[k] = t.Format(time.RFC3339Nano)
+		case uuid.UUID:
+			completeMap[typeKey] = OdataGUID
+			completeMap[k] = t.String()
+		case int64:
+			completeMap[typeKey] = OdataInt64
+			completeMap[k] = fmt.Sprintf("%v", v)
+		default:
+			completeMap[k] = v
+		}
+		if strings.HasSuffix(k, OdataTypeSuffix) {
+			if !(completeMap[k] == OdataBinary ||
+				completeMap[k] == OdataDateTime ||
+				completeMap[k] == OdataGUID ||
+				completeMap[k] == OdataInt64) {
+				return nil, fmt.Errorf("Odata.type annotation %v value is not valid", k)
+			}
+			valueKey := strings.TrimSuffix(k, OdataTypeSuffix)
+			if _, ok := completeMap[valueKey]; !ok {
+				return nil, fmt.Errorf("Odata.type annotation %v defined without value defined", k)
+			}
+		}
+	}
+	return json.Marshal(completeMap)
+}
+
+// UnmarshalJSON is a custom unmarshaller for entities
+func (e *Entity) UnmarshalJSON(data []byte) error {
+	errorTemplate := "Deserializing error: %v"
+
+	props := map[string]interface{}{}
+	err := json.Unmarshal(data, &props)
+	if err != nil {
+		return err
+	}
+
+	// deselialize metadata
+	props = assignFromMap(props, &e.OdataMetadata, "odata.metadata")
+	props = assignFromMap(props, &e.OdataType, "odata.type")
+	props = assignFromMap(props, &e.OdataID, "odata.id")
+	props = assignFromMap(props, &e.OdataEtag, "odata.etag")
+	props = assignFromMap(props, &e.OdataEditLink, "odata.editLink")
+	props = assignFromMap(props, &e.PartitionKey, partitionKeyNode)
+	props = assignFromMap(props, &e.RowKey, rowKeyNode)
+
+	// deserialize timestamp
+	str, ok := props["Timestamp"].(string)
+	if !ok {
+		return fmt.Errorf(errorTemplate, "Timestamp casting error")
+	}
+	t, err := time.Parse(time.RFC3339Nano, str)
+	if err != nil {
+		return fmt.Errorf(errorTemplate, err)
+	}
+	e.TimeStamp = t
+	delete(props, "Timestamp")
+	delete(props, "Timestamp@odata.type")
```

Comment:
It looks like there is a const declared below for `"@odata.type"` Perhaps we should clean this up to something like:

``` Go
const timestamp = "Timestamp" // There's probably a better name and place for this const
delete(props, timstamp)
delete(props, timestamp + OdataTypeSuffix)
```